File: 297-serialize-and-deserialize-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        preorder(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return buildTree(in);
    }

private:
    void preorder(TreeNode * root, ostringstream & out) {
        if (root) {
            out << root->val << " ";
            preorder(root->left, out);
            preorder(root->right, out);
        } else out << "# ";
    }
    
    TreeNode * buildTree(istringstream & in) {
        string val;
        in >> val;
        if (val == "#") return nullptr;
        TreeNode *root = new TreeNode(stoi(val));
        root->left = buildTree(in);
        root->right = buildTree(in);
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: 297-serialize-and-deserialize-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        self.res = []
        def preorder(root):
            if root is None:
                self.res.append('#')
            else:
                self.res.append(str(root.val))
                preorder(root.left)
                preorder(root.right)
        preorder(root)
        return ','.join(self.res)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        data = data.split(',')
        def build(data):
            if not data: return None

            val = data.pop(0)
            if val == '#': return None
            root = TreeNode(int(val))
            root.left = build(data)
            root.right = build(data)
            return root
        return build(data)

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))